Native Tests fail when quarkus.http.test-port is set to 0

**Describe the bug**

When running a native test with `quarkus.http.test-port=0` you get the following :

```
[INFO] Running org.acme.NativeExampleResourceIT
Executing target/code-with-quarkus-1.0.0-SNAPSHOT-runner, -Dquarkus.http.port=0, -Dquarkus.http.ssl-port=8444, -Dtest.url=http://localhost:0
```

The runner is executed with `localhost:0` (instead of a random port). So it hangs, and then the test fails with:

```
Caused by: java.lang.RuntimeException: Unable to start native image in 60s
```

**Expected behavior**

The test should pass using a random port (like JVM mode tests).

**Actual behavior**


**To Reproduce**
Steps to reproduce the behavior:
1. Go to code.quarkus.io and generate a new project
2. `mvn verify -Pnative`  works
3. Add the property `quarkus.http.test-port=0`to the `application.properties` file
4. `mvn verify -Pnative`  FAILS

**Configuration**
```properties
quarkus.http.test-port=0
```

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: Darwin iMac-Pro-de-Antonio.local 19.6.0 Darwin Kernel Version 19.6.0: Thu Jun 18 20:49:00 PDT 2020; root:xnu-6153.141.1~1/RELEASE_X86_64 x86_64
 - Output of `java -version`: java version "11.0.6" 2020-01-14 LTS
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 1.7.1-FINAL
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 3.6.3_1

**Additional context**
(Add any other context about the problem here.)

--- a/integration-tests/spring-web/src/main/resources/application.properties
+++ b/integration-tests/spring-web/src/main/resources/application.properties
@@ -2,7 +2,8 @@ quarkus.security.users.file.enabled=true
 quarkus.security.users.file.users=test-users.properties
 quarkus.security.users.file.roles=test-roles.properties
 quarkus.security.users.file.plain-text=true
-%test.quarkus.http.test-port=0
+quarkus.http.test-port=0
+quarkus.http.port=0
 
 # we add this because we also use @TestSecurity which means that basic auth is disabled by default because @TestSecurity
 # contributes TestHttpAuthenticationMechanism

--- a/test-framework/common/src/main/java/io/quarkus/test/common/NativeImageLauncher.java
+++ b/test-framework/common/src/main/java/io/quarkus/test/common/NativeImageLauncher.java
@@ -18,9 +18,14 @@ import java.util.Map;
 import java.util.OptionalInt;
 import java.util.OptionalLong;
 import java.util.ServiceLoader;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 import org.eclipse.microprofile.config.Config;
 import org.eclipse.microprofile.config.spi.ConfigProviderResolver;
+import org.wildfly.common.lock.Locks;
 
 import io.quarkus.runtime.configuration.ConfigUtils;
 import io.quarkus.runtime.configuration.QuarkusConfigFactory;
@@ -97,6 +102,8 @@ public class NativeImageLauncher implements Closeable {
         args.add(path);
         args.add("-Dquarkus.http.port=" + port);
         args.add("-Dquarkus.http.ssl-port=" + httpsPort);
+        // this won't be correct when using the random port but it's really only used by us for the rest client tests
+        // in the main module, since those tests hit the application itself
         args.add("-Dtest.url=" + TestHTTPResourceManager.getUri());
         args.add("-Dquarkus.log.file.path=" + PropertyTestUtil.getLogFileLocation());
         if (profile != null) {
@@ -109,10 +116,37 @@ public class NativeImageLauncher implements Closeable {
         System.out.println("Executing " + args);
 
         quarkusProcess = Runtime.getRuntime().exec(args.toArray(new String[args.size()]));
-        new Thread(new ProcessReader(quarkusProcess.getInputStream())).start();
+
+        PortCapturingProcessReader portCapturingProcessReader = null;
+        if (port == 0) {
+            // when the port is 0, then the application starts on a random port and the only way for us to figure it out
+            // is to capture the output
+            portCapturingProcessReader = new PortCapturingProcessReader(quarkusProcess.getInputStream());
+        }
+        new Thread(portCapturingProcessReader != null ? portCapturingProcessReader
+                : new ProcessReader(quarkusProcess.getInputStream())).start();
         new Thread(new ProcessReader(quarkusProcess.getErrorStream())).start();
 
-        waitForQuarkus();
+        if (portCapturingProcessReader != null) {
+            try {
+                portCapturingProcessReader.awaitForPort();
+            } catch (InterruptedException ignored) {
+
+            }
+            if (portCapturingProcessReader.port == null) {
+                quarkusProcess.destroy();
+                throw new RuntimeException("Unable to determine actual running port as dynamic port was used");
+            }
+
+            waitForQuarkus(portCapturingProcessReader.port);
+
+            System.setProperty("quarkus.http.port", portCapturingProcessReader.port.toString()); //set the port as a system property in order to have it applied to Config
+            System.setProperty("quarkus.http.test-port", portCapturingProcessReader.port.toString()); // needed for RestAssuredManager
+            installAndGetSomeConfig(); // reinitialize the configuration to make sure the actual port is used
+            System.setProperty("test.url", TestHTTPResourceManager.getUri());
+        } else {
+            waitForQuarkus(port);
+        }
     }
 
     private static String guessPath(Class<?> testClass) {
@@ -198,7 +232,7 @@ public class NativeImageLauncher implements Closeable {
         System.err.println("======================================================================================");
     }
 
-    private void waitForQuarkus() {
+    private void waitForQuarkus(int port) {
         long bailout = System.currentTimeMillis() + imageWaitTime * 1000;
 
         while (System.currentTimeMillis() < bailout) {
@@ -220,6 +254,7 @@ public class NativeImageLauncher implements Closeable {
             }
         }
 
+        quarkusProcess.destroy();
         throw new RuntimeException("Unable to start native image in " + imageWaitTime + "s");
     }
 
@@ -227,7 +262,7 @@ public class NativeImageLauncher implements Closeable {
         this.systemProps.putAll(systemProps);
     }
 
-    private static final class ProcessReader implements Runnable {
+    private static class ProcessReader implements Runnable {
 
         private final InputStream inputStream;
 
@@ -237,14 +272,92 @@ public class NativeImageLauncher implements Closeable {
 
         @Override
         public void run() {
+            handleStart();
             byte[] b = new byte[100];
             int i;
             try {
                 while ((i = inputStream.read(b)) > 0) {
-                    System.out.print(new String(b, 0, i, StandardCharsets.UTF_8));
+                    String str = new String(b, 0, i, StandardCharsets.UTF_8);
+                    System.out.print(str);
+                    handleString(str);
                 }
             } catch (IOException e) {
-                //ignore
+                handleError(e);
+            }
+        }
+
+        protected void handleStart() {
+
+        }
+
+        protected void handleString(String str) {
+
+        }
+
+        protected void handleError(IOException e) {
+
+        }
+    }
+
+    private static final class PortCapturingProcessReader extends ProcessReader {
+        private Integer port;
+
+        private boolean portDetermined = false;
+        private StringBuilder sb = new StringBuilder();
+        private final Lock lock = Locks.reentrantLock();
+        private final Condition portDeterminedCondition = lock.newCondition();
+        private final Pattern portRegex = Pattern.compile("Listening on:\\s+https?://.*:(\\d+)");
+
+        private PortCapturingProcessReader(InputStream inputStream) {
+            super(inputStream);
+        }
+
+        @Override
+        protected void handleStart() {
+            lock.lock();
+        }
+
+        @Override
+        protected void handleString(String str) {
+            if (portDetermined) { // we are done with determining the port
+                return;
+            }
+            sb.append(str);
+            String currentOutput = sb.toString();
+            Matcher regexMatcher = portRegex.matcher(currentOutput);
+            if (!regexMatcher.find()) { // haven't read enough data yet
+                if (currentOutput.contains("Exception")) {
+                    portDetermined(null);
+                }
+                return;
+            }
+            portDetermined(Integer.valueOf(regexMatcher.group(1)));
+        }
+
+        private void portDetermined(Integer portValue) {
+            this.port = portValue;
+            try {
+                portDetermined = true;
+                sb = null;
+                portDeterminedCondition.signal();
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        @Override
+        protected void handleError(IOException e) {
+            portDetermined(null);
+        }
+
+        public void awaitForPort() throws InterruptedException {
+            lock.lock();
+            try {
+                while (!portDetermined) {
+                    portDeterminedCondition.await();
+                }
+            } finally {
+                lock.unlock();
             }
         }
     }

